modules/transactions/src/main/java/org/apache/ignite/internal/tx/LockMode.java

```diff
@@ -21,9 +21,121 @@
  * Lock mode.
  */
 public enum LockMode {
-    SHARED,
-    EXCLUSIVE,
     INTENTION_SHARED,
     INTENTION_EXCLUSIVE,
-    SHARED_AND_INTENTION_EXCLUSIVE
+    SHARED,
+    SHARED_AND_INTENTION_EXCLUSIVE,
+    EXCLUSIVE;
+
+    /**
+     * Is this lock mode compatible with the specified lock mode.
+     *
+     * @param lockMode Lock mode.
+     * @return Is this lock mode compatible with the specified lock mode.
+     */
+    public boolean isCompatible(LockMode lockMode) {
```

Comment:
Let's steal Alexey's compatibility and upgrade matrix

```
    protected static final boolean[][] COMPAT_MATRIX = {
            {true, true, true, true, false},
            {true, true, false, false, false},
            {true, false, true, false, false},
            {true, false, false, false, false},
            {false, false, false, false, false},
    };

    protected static final LockMode[][] UPGRADE_MATRIX = {
            {LockMode.IS, LockMode.IX, LockMode.S, LockMode.SIX, LockMode.X},
            {LockMode.IX, LockMode.IX, LockMode.SIX, LockMode.SIX, LockMode.X},
            {LockMode.S, LockMode.SIX, LockMode.S, LockMode.SIX, LockMode.X},
            {LockMode.SIX, LockMode.SIX, LockMode.SIX, LockMode.SIX, LockMode.X},
            {LockMode.X, LockMode.X, LockMode.X, LockMode.X, LockMode.X},
    };
```

Reply:
Fixed